doc/source/reference/arrays.promotion.rst

```diff
@@ -0,0 +1,263 @@
+while following the (green) lines.
+
+Note the following specific rules and observations:
+1. When a Python ``float`` or ``complex`` interacts with a NumPy integer
+   the result will be ``float64`` or ``complex128`` (yellow border).
+   NumPy booleans will also be cast to the default integer.[#default-int]
+   This is not relevant when additionally NumPy floating point values are
+   involved.
+2. The precision is drawn such that ``float16 < int16 < uint16`` because
+   large ``uint16`` do not fit ``int16`` and large ``int16`` will lose precision
+   when stored in a ``float16``.
+   This pattern however is broken since NumPy always considers ``float64``
+   and ``complex128`` to be acceptable promotion results for any integer
+   value.
+3. A special case the above leads to is that NumPy will promote many
+   combinations of signed and unsigned integers to ``float64`` because no
+   integer dtype can hold both inputs.
+   This can unfortunately be 
+
+The precision here comes from the bit size of the numerical value but
+an ``int32`` cannot always be stored in a ``float32`` without loss of
+precision, this leads to the 
+
+
+Notable or special promotion behaviors
+--------------------------------------
+
+In NumPy promotion refers to what specific functions do with the result and
+in some cases, this means that NumPy may deviate from what the `np.result_type`
+would give.
+
+Behavior of ``sum`` and ``prod``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+**``np.sum`` and ``np.prod``:** Will alway return the default integer type
+when summing over integer values (or booleans).  This is usually an ``int64``.
+The reason for this is that integer summations are otherwise very likely
+to overflow and give confusing results.
+This rule also applies to the underlying ``np.add.reduce``, ``np.multiply.reduce``
+and other reduction methods.
+
+Notable behavior with NumPy or Python integer scalars
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+NumPy promotion refers to the result dtype and operation precision,
+but the operation will sometimes dictate that result.
+Division always returns floating point values and comparison always booleans.
+
+This leads to some special behaviors which may appear as "exceptions" to
+the rules:
+* NumPy comparisons with Python integers or mixed precision integers always
+  return the correct result.  The inputs will never be cast in a way which
+  loses precision.
+* Equality comparisons between types which cannot be promoted will be
+  considered all ``False`` (equality) or all ``True`` (not-equal).
+* Unary math functions like ``np.sin`` that always return floating point
+  values, accept any Python integer input by converting it to ``float64``.
+* Division always returns floating point values and thus also allows divisions
+  between any NumPy integer with any Python integer value by casting both
+  to ``float64``.
+
+It should be noted that while these exceptions can apply to more functions,
```

Comment:
I would phrase this in a friendly way as,
```
In principle, some of these exceptions may make sense for other functions.
Please raise an issue if you feel this is the case.
```
(eg., really `np.clip` is another case where we may want exceptions...)

Reply:
adopted